src/box/txn.c

```diff
@@ -652,6 +652,8 @@ txn_commit_stmt(struct txn *txn, struct request *request)
 			goto fail;
 		assert(stmt->row != NULL);
 		txn_update_row_counts(txn, stmt, 1);
+	} else {
+		txn->has_temporary_stmts = true;
```

Comment:
That unfortunately won't work well if the statement is rolled back by a savepoint-rollback. The flag would remain set but inaccurate. I would suggest to rather walk the statements in `txn_is_fully_remote()` to check if any of them are temporary. It is linear complexity, but is needed only for true DDL (i.e. not on temporary spaces), which can only be rare. Given that it aborts all concurrent txns (excluding applier and temporary spaces).

Also the flag isn't initialized now in `txn_begin`. Can as well become `true` randomly and start failing. And for flags anyway we use `enum txn_flag`. But I suggest not to add it at all really. It is deductible from the statements and is needed only on rollback (perf isn't important).

Reply:
> I would suggest to rather walk the statements in `txn_is_fully_remote()` to check if any of them are temporary.

I don't understand how you suggest to check if they are temporary. Temporary and nop statements are indistinguishable from this point in the code. Only if we add a `temporary` flag to `txn_stmt`.

> That unfortunately won't work well if the statement is rolled back by a savepoint-rollback. The flag would remain set but inaccurate.

Is this a common case? If not, I would suggest leaving it as is, since we are trying to fix a common source of pain for Picodata user workloads here.

> Also the flag isn't initialized now in `txn_begin`.

Fixed, thanks.

> And for flags anyway we use `enum txn_flag`.

Not always, for instance we have a `is_schema_changed` flag. I think this is a temporary solution, and it should also be just a separate flag.

Reply:
Fixed by applying your suggestion, thanks for the catch! Added test cases for savepoint-rollback.